youtube_dl/extractor/frontendmaster.py

```diff
@@ -0,0 +1,271 @@
+    _NETRC_MACHINE = 'frontend-masters'
+    _TEST = {
+        'url': 'https://frontendmasters.com/courses/content-strategy/introduction/',
+        'md5': '5f176d4f170778524f40a06307a929f6',
+        'info_dict': {
+            'id': 'introduction',
+            'title': 'Introduction',
+            'display_id': 'content-strategy',
+            'ext': 'mp4'
+        },
+        'skip': 'Requires FrontendMasters account credentials'
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        course_id = self._match_course_id(url)
+        course_json_content = self._download_course(course_id=course_id,
+                                                    url=url,
+                                                    display_id=course_id)
+
+        lesson_index = course_json_content.get('lessonSlugs').index(video_id)
+        lesson_hash = course_json_content.get('lessonHashes')[lesson_index]
+        lesson_section_elements = course_json_content.get('lessonElements')
+        lesson_data = course_json_content.get('lessonData')[lesson_hash]
+        lesson_source_base = lesson_data.get('sourceBase')
+        course_sections_pairing = self._pair_section_with_video_elemen_index(lesson_section_elements)
+
+        lesson_title = lesson_data.get('title')
+        lesson_description = lesson_data.get('description')
+        lesson_index = lesson_data.get('index')
+        lesson_slug = lesson_data.get('slug')
+        lesson_thumbnail_url = lesson_data.get('thumbnail')
+        lesson_section = course_sections_pairing.get(lesson_index)[0]
+        lesson_section_number = course_sections_pairing.get(lesson_index)[1]
+
+
+        QUALITIES_PREFERENCE = ('low', 'medium', 'high')
+        quality_key = qualities(QUALITIES_PREFERENCE)
+        QUALITIES = {
+            'low': {'width': 480, 'height': 360},
+            'medium': {'width': 1280, 'height': 720},
+            'high': {'width': 1920, 'height': 1080}
+        }
+
+        AllowedQuality = collections.namedtuple('AllowedQuality', ['ext', 'qualities'])
+        ALLOWED_QUALITIES = [
+            AllowedQuality('webm', ['low', 'medium', 'high']),
+            AllowedQuality('mp4', ['low', 'medium', 'high'])
+        ]
+
+        cookies = self._get_cookies(self._COOKIES_BASE)
+        cookies_str = ";".join(["%s=%s" % (cookie.key, cookie.value)
+                                for cookie in cookies.values()])
+        video_request_url = "%s/source"
+        video_request_headers = {
+            "origin": "https://frontendmasters.com",
+            "referer": lesson_source_base,
+            "cookie": cookies_str,
+            'user-agent': "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) "
+                          "AppleWebKit/537.36 (KHTML, like Gecko) "
+                          "Chrome/66.0.3359.117 Safari/537.36"
+        }
+
+        if self._downloader.params.get('listformats', False):
```

Comment:
Remove.

Reply:
Why? I took inspiration from the Pluralsight Extractor for the Qualities bit, and for what concern the cookies etc. it's to avoid "400 - Bad request" response from the server